jiant/trainer.py

```diff
@@ -591,31 +605,29 @@ def clip_function(grad):
             n_batches_since_val = task_info["n_batches_since_val"]
             tr_loss = task_info["loss"]
 
-            for batch in itertools.islice(tr_generator, 1):
-                n_batches_since_val += 1
-                total_batches_trained += 1
-                optimizer.zero_grad()
+            for batch in itertools.islice(tr_generator, self._accumulation_steps):
                 output_dict = self._forward(batch, task=task)
-                assert_for_log(
-                    "loss" in output_dict, "Model must return a dict containing a 'loss' key"
-                )
-                loss = get_output_attribute(
-                    output_dict, "loss", self._cuda_device
-                )  # optionally scale loss
+                assert_for_log("loss" in output_dict, "Model must return a dict with 'loss' key")
+                loss = get_output_attribute(output_dict, "loss", self._cuda_device)
+                if self._accumulation_steps > 1:
+                    loss = loss / self._accumulation_steps
                 loss *= scaling_weights[task.name]
                 loss.backward()
                 assert_for_log(not torch.isnan(loss).any(), "NaNs in loss.")
                 tr_loss += loss.data.cpu().numpy()
+            n_batches_since_val += 1
+            total_batches_trained += 1
+            n_step += 1
```

Comment:
This doesn't seem to match what we'd talked about. I thought you'd said that we increment the step counter every time we run the model, not every time we take a gradient step. Either way is fine, but make sure we're consistent within the PR.

Reply:
With 9995fe5 we count `batches` and `steps` separately (incrementing `steps` following each optimizer step, and `batches` on each update). I think this should resolve the batch/step counting confusion.